ffi/src/sspi/sec_winnt_auth_identity.rs

```diff
@@ -718,17 +719,39 @@ pub unsafe extern "system" fn SspiEncodeStringsAsAuthIdentity(
         check_null!(psz_domain_name);
         check_null!(psz_packed_credentials_string);
 
-        let user = c_w_str_to_string(psz_user_name);
-        let domain = c_w_str_to_string(psz_domain_name);
-        let password = c_w_str_to_string(psz_packed_credentials_string);
+        let user_length = w_str_len(psz_user_name);
+        let domain_length = w_str_len(psz_domain_name);
+        let password_length = w_str_len(psz_packed_credentials_string);
+
+        if user_length == 0 || domain_length == 0 || password_length == 0 {
+            return ErrorKind::InvalidParameter.to_u32().unwrap();
+        }
+
+        let user = unsafe { libc::malloc(user_length * 2) as *mut SecWChar };
+        if user.is_null() {
+            return ErrorKind::InternalError.to_u32().unwrap();
+        }
+        copy_nonoverlapping(psz_user_name, user, user_length);
+
+        let domain = unsafe { libc::malloc(domain_length * 2) as *mut SecWChar };
+        if domain.is_null() {
+            return ErrorKind::InternalError.to_u32().unwrap();
+        }
+        copy_nonoverlapping(psz_domain_name, domain, domain_length);
+
+        let password = unsafe { libc::malloc(password_length * 2) as *mut SecWChar };
+        if password.is_null() {
+            return ErrorKind::InternalError.to_u32().unwrap();
+        }
+        copy_nonoverlapping(psz_packed_credentials_string, password, password_length);
 
         let auth_identity = SecWinntAuthIdentityW {
-            user: psz_user_name,
-            user_length: user.len().try_into().unwrap(),
-            domain: psz_domain_name,
-            domain_length: domain.len().try_into().unwrap(),
-            password: psz_packed_credentials_string,
-            password_length: password.len().try_into().unwrap(),
+            user,
+            user_length: user_length.try_into().unwrap(),
+            domain,
+            domain_length: domain_length.try_into().unwrap(),
+            password,
+            password_length: password_length.try_into().unwrap(),
```

Comment:
we shouldn't use provided buffers (because we don't manage their lifetime) here but copy the data and then use it